Use object lifetimes to manage message receiver registrations

achristensen07 opened the pull request:
#### a2878138acd817eaa41c98ea75b1bc8e2e830729
<pre>
Use object lifetimes to manage message receiver registrations
<a href="https://bugs.webkit.org/show_bug.cgi?id=258829">https://bugs.webkit.org/show_bug.cgi?id=258829</a>
rdar://111064194

Reviewed by J Pascoe.

The receiving of messages to Messages::WebPageProxy::messageReceiverName
is a tricky thing.  When a RemotePageProxy is created during
ProvisionalPageProxy::didCommitLoadForFrame when the frame is the result of
a cross-site window.open in another process (enabled by an off-by-default
feature flag) then we need to transfer the receiving of the messages from
the WebPageProxy to the RemotePageProxy.  We then don&apos;t want the WebPageProxy
to start and stop messages from being received for that name any more.
To do this in the most straightforward way I could think of, I tied the
registrations to the lifetime of a new object WebPageProxyMessageReceiverRegistration
which also has the ability to start/stop/transfer its message receiving.
It&apos;s a little like a smart pointer, but with lots of metadata for each operation.
It removes the need for explicit calls to remoteMessageReceiver with
Messages::WebPageProxy::messageReceiverName.

I did something quite similar for the VisitedLinkStore registration of a
RemotePageProxy, which only needs to register and unregister when
RemotePageProxy::injectPageIntoNewProcess is called, which happens if that
RemotePageProxy is used for a remote iframe but not for a cross-site window.open.

I also introduce RemotePageDrawingAreaProxy which is similar for the DrawingAreaProxy
messages.  Its existence makes it so that each DrawingArea has one unique proxy
object it communicates with, which will aid in future development of drawing
with site isolation.  Right now it just forwards all its messages to the DrawingAreaProxy
of the page, but in the future it can have the ability to intercept those messages.
In order to know what message receiver names to register for, I replaced the virtual calls
in RemoteLayerTreeDrawingAreaProxy::startReceivingMessages with a function that returns
a std::span of names.  This is safe because all implementations return pointers to
static std::array&apos;s, so their lifetimes are forever.  This saves unnecessary allocations.

A few other changes are needed to fix the build after adding new files.
Aren&apos;t unified builds fun?

* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/DrawingAreaProxy.cpp:
(WebKit::DrawingAreaProxy::startReceivingMessages):
(WebKit::DrawingAreaProxy::stopReceivingMessages):
(WebKit::DrawingAreaProxy::messageReceiverNames const):
* Source/WebKit/UIProcess/DrawingAreaProxy.h:
* Source/WebKit/UIProcess/PerActivityStateCPUUsageSampler.cpp:
* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::ProvisionalPageProxy):
(WebKit::ProvisionalPageProxy::~ProvisionalPageProxy):
(WebKit::ProvisionalPageProxy::didCommitLoadForFrame):
* Source/WebKit/UIProcess/ProvisionalPageProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::messageReceiverNames const):
(WebKit::RemoteLayerTreeDrawingAreaProxy::startReceivingMessages): Deleted.
(WebKit::RemoteLayerTreeDrawingAreaProxy::stopReceivingMessages): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeEventDispatcher.h:
* Source/WebKit/UIProcess/RemotePageDrawingAreaProxy.cpp: Added.
(WebKit::RemotePageDrawingAreaProxy::RemotePageDrawingAreaProxy):
(WebKit::RemotePageDrawingAreaProxy::~RemotePageDrawingAreaProxy):
(WebKit::RemotePageDrawingAreaProxy::didReceiveMessage):
(WebKit::RemotePageDrawingAreaProxy::didReceiveSyncMessage):
* Source/WebKit/UIProcess/RemotePageDrawingAreaProxy.h: Added.
* Source/WebKit/UIProcess/RemotePageProxy.cpp:
(WebKit::RemotePageProxy::RemotePageProxy):
(WebKit::RemotePageProxy::injectPageIntoNewProcess):
(WebKit::RemotePageProxy::~RemotePageProxy):
* Source/WebKit/UIProcess/RemotePageProxy.h:
(WebKit::RemotePageProxy::create):
(WebKit::RemotePageProxy::messageReceiverRegistration):
* Source/WebKit/UIProcess/RemotePageVisitedLinkStoreRegistration.h: Added.
(WebKit::RemotePageVisitedLinkStoreRegistration::RemotePageVisitedLinkStoreRegistration):
(WebKit::RemotePageVisitedLinkStoreRegistration::~RemotePageVisitedLinkStoreRegistration):
* Source/WebKit/UIProcess/SuspendedPageProxy.cpp:
(WebKit::SuspendedPageProxy::SuspendedPageProxy):
(WebKit::SuspendedPageProxy::~SuspendedPageProxy):
(WebKit::SuspendedPageProxy::didProcessRequestToSuspend):
* Source/WebKit/UIProcess/SuspendedPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::~WebPageProxy):
(WebKit::WebPageProxy::addAllMessageReceivers):
(WebKit::WebPageProxy::removeAllMessageReceivers):
(WebKit::WebPageProxy::messageReceiverRegistration):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxyInternals.h:
* Source/WebKit/UIProcess/WebPageProxyMessageReceiverRegistration.cpp: Added.
(WebKit::WebPageProxyMessageReceiverRegistration::~WebPageProxyMessageReceiverRegistration):
(WebKit::WebPageProxyMessageReceiverRegistration::startReceivingMessages):
(WebKit::WebPageProxyMessageReceiverRegistration::stopReceivingMessages):
(WebKit::WebPageProxyMessageReceiverRegistration::transferMessageReceivingFrom):
* Source/WebKit/UIProcess/WebPageProxyMessageReceiverRegistration.h: Added.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ProcessSwapOnNavigation.mm:

Canonical link: <a href="https://commits.webkit.org/265771@main">https://commits.webkit.org/265771@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fb84c7a9e10e3bc0f2fad00363b6bafc13dd7fad

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/11773 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/11972 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/12340 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/13417 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/11212 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/11792 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/14361 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/11953 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14074 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/11937 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/12772 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/9988 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/13838 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10064 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/10686 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/17818 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11141 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/10841 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/14013 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/11249 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/9291 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10423 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/14706 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1327 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/11105 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pascoej reviewed (approved):
This looks great. Will `SameSiteWindowWithOpenerNavigateToFile` pass now?

achristensen07 committed:
Use object lifetimes to manage message receiver registrations
https://bugs.webkit.org/show_bug.cgi?id=258829
rdar://111064194

Reviewed by J Pascoe.

The receiving of messages to Messages::WebPageProxy::messageReceiverName
is a tricky thing.  When a RemotePageProxy is created during
ProvisionalPageProxy::didCommitLoadForFrame when the frame is the result of
a cross-site window.open in another process (enabled by an off-by-default
feature flag) then we need to transfer the receiving of the messages from
the WebPageProxy to the RemotePageProxy.  We then don't want the WebPageProxy
to start and stop messages from being received for that name any more.
To do this in the most straightforward way I could think of, I tied the
registrations to the lifetime of a new object WebPageProxyMessageReceiverRegistration
which also has the ability to start/stop/transfer its message receiving.
It's a little like a smart pointer, but with lots of metadata for each operation.
It removes the need for explicit calls to remoteMessageReceiver with
Messages::WebPageProxy::messageReceiverName.

I did something quite similar for the VisitedLinkStore registration of a
RemotePageProxy, which only needs to register and unregister when
RemotePageProxy::injectPageIntoNewProcess is called, which happens if that
RemotePageProxy is used for a remote iframe but not for a cross-site window.open.

I also introduce RemotePageDrawingAreaProxy which is similar for the DrawingAreaProxy
messages.  Its existence makes it so that each DrawingArea has one unique proxy
object it communicates with, which will aid in future development of drawing
with site isolation.  Right now it just forwards all its messages to the DrawingAreaProxy
of the page, but in the future it can have the ability to intercept those messages.
In order to know what message receiver names to register for, I replaced the virtual calls
in RemoteLayerTreeDrawingAreaProxy::startReceivingMessages with a function that returns
a std::span of names.  This is safe because all implementations return pointers to
static std::array's, so their lifetimes are forever.  This saves unnecessary allocations.

A few other changes are needed to fix the build after adding new files.
Aren't unified builds fun?

* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/DrawingAreaProxy.cpp:
(WebKit::DrawingAreaProxy::startReceivingMessages):
(WebKit::DrawingAreaProxy::stopReceivingMessages):
(WebKit::DrawingAreaProxy::messageReceiverNames const):
* Source/WebKit/UIProcess/DrawingAreaProxy.h:
* Source/WebKit/UIProcess/PerActivityStateCPUUsageSampler.cpp:
* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::ProvisionalPageProxy):
(WebKit::ProvisionalPageProxy::~ProvisionalPageProxy):
(WebKit::ProvisionalPageProxy::didCommitLoadForFrame):
* Source/WebKit/UIProcess/ProvisionalPageProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::messageReceiverNames const):
(WebKit::RemoteLayerTreeDrawingAreaProxy::startReceivingMessages): Deleted.
(WebKit::RemoteLayerTreeDrawingAreaProxy::stopReceivingMessages): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeEventDispatcher.h:
* Source/WebKit/UIProcess/RemotePageDrawingAreaProxy.cpp: Added.
(WebKit::RemotePageDrawingAreaProxy::RemotePageDrawingAreaProxy):
(WebKit::RemotePageDrawingAreaProxy::~RemotePageDrawingAreaProxy):
(WebKit::RemotePageDrawingAreaProxy::didReceiveMessage):
(WebKit::RemotePageDrawingAreaProxy::didReceiveSyncMessage):
* Source/WebKit/UIProcess/RemotePageDrawingAreaProxy.h: Added.
* Source/WebKit/UIProcess/RemotePageProxy.cpp:
(WebKit::RemotePageProxy::RemotePageProxy):
(WebKit::RemotePageProxy::injectPageIntoNewProcess):
(WebKit::RemotePageProxy::~RemotePageProxy):
* Source/WebKit/UIProcess/RemotePageProxy.h:
(WebKit::RemotePageProxy::create):
(WebKit::RemotePageProxy::messageReceiverRegistration):
* Source/WebKit/UIProcess/RemotePageVisitedLinkStoreRegistration.h: Added.
(WebKit::RemotePageVisitedLinkStoreRegistration::RemotePageVisitedLinkStoreRegistration):
(WebKit::RemotePageVisitedLinkStoreRegistration::~RemotePageVisitedLinkStoreRegistration):
* Source/WebKit/UIProcess/SuspendedPageProxy.cpp:
(WebKit::SuspendedPageProxy::SuspendedPageProxy):
(WebKit::SuspendedPageProxy::~SuspendedPageProxy):
(WebKit::SuspendedPageProxy::didProcessRequestToSuspend):
* Source/WebKit/UIProcess/SuspendedPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::~WebPageProxy):
(WebKit::WebPageProxy::addAllMessageReceivers):
(WebKit::WebPageProxy::removeAllMessageReceivers):
(WebKit::WebPageProxy::messageReceiverRegistration):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxyInternals.h:
* Source/WebKit/UIProcess/WebPageProxyMessageReceiverRegistration.cpp: Added.
(WebKit::WebPageProxyMessageReceiverRegistration::~WebPageProxyMessageReceiverRegistration):
(WebKit::WebPageProxyMessageReceiverRegistration::startReceivingMessages):
(WebKit::WebPageProxyMessageReceiverRegistration::stopReceivingMessages):
(WebKit::WebPageProxyMessageReceiverRegistration::transferMessageReceivingFrom):
* Source/WebKit/UIProcess/WebPageProxyMessageReceiverRegistration.h: Added.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ProcessSwapOnNavigation.mm:

Canonical link: https://commits.webkit.org/265771@main
